File: apps/tauri/src/commands/cash_interest.rs
use std::sync::Arc;

use crate::context::ServiceContext;
use log::debug;
use tauri::State;
use wealthfolio_core::interest::{CashInterestSyncResult, Mp2DividendRates};

#[tauri::command]
pub async fn sync_cash_interest(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<CashInterestSyncResult, String> {
    debug!("Syncing cash interest accrual...");
    state
        .interest_accrual_service()
        .sync()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn sync_cash_interest_account(
    account_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<CashInterestSyncResult, String> {
    debug!("Syncing cash interest for account {account_id}...");
    state
        .interest_accrual_service()
        .sync_account(&account_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_mp2_rates(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Mp2DividendRates, String> {
    debug!("Fetching MP2 dividend rates...");
    state
        .interest_accrual_service()
        .get_mp2_rates()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_mp2_rates(
    rates: Mp2DividendRates,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Mp2DividendRates, String> {
    debug!("Updating MP2 dividend rates...");
    state
        .interest_accrual_service()
        .set_mp2_rates(&rates)
        .await
        .map_err(|e| e.to_string())?;
    Ok(rates)
}

#[tauri::command]
pub async fn remove_auto_interest(state: State<'_, Arc<ServiceContext>>) -> Result<usize, String> {
    debug!("Removing auto-created interest activities...");
    state
        .interest_accrual_service()
        .remove_auto_created()
        .await
        .map_err(|e| e.to_string())
}

File: apps/frontend/src/adapters/shared/cash-interest.ts
import { invoke, logger } from "./platform";

export interface CashInterestSyncAccountResult {
  accountId: string;
  accountName: string;
  created: number;
  amended: number;
  removed: number;
  skippedOverrides: number;
  skippedDuplicates: number;
}

export interface CashInterestSyncResult {
  created: number;
  amended: number;
  removed: number;
  skipped: number;
  skippedOverrides: number;
  skippedDuplicates: number;
  accounts: CashInterestSyncAccountResult[];
  errors: string[];
  netCashAdded: string;
}

/**
 * Pag-IBIG declares one MP2 rate per year for all members, so these are app-wide
 * rather than per-account. Keys are the calendar year the dividend was earned.
 */
export interface Mp2DividendRates {
  rates: Record<string, number>;
}

export const syncCashInterest = async (): Promise<CashInterestSyncResult> => {
  try {
    return await invoke<CashInterestSyncResult>("sync_cash_interest");
  } catch (error) {
    logger.error("Error syncing cash interest.");
    throw error;
  }
};

export const syncCashInterestAccount = async (
  accountId: string,
): Promise<CashInterestSyncResult> => {
  try {
    return await invoke<CashInterestSyncResult>("sync_cash_interest_account", { accountId });
  } catch (error) {
    logger.error("Error syncing cash interest for account.");
    throw error;
  }
};

export const removeAutoInterest = async (): Promise<number> => {
  try {
    return await invoke<number>("remove_auto_interest");
  } catch (error) {
    logger.error("Error removing auto interest.");
    throw error;
  }
};

export const removeAutoInterestAccount = async (accountId: string): Promise<number> => {
  try {
    return await invoke<number>("remove_auto_interest_account", { accountId });
  } catch (error) {
    logger.error("Error removing auto interest for account.");
    throw error;
  }
};

export const getMp2Rates = async (): Promise<Mp2DividendRates> => {
  try {
    return await invoke<Mp2DividendRates>("get_mp2_rates");
  } catch (error) {
    logger.error("Error fetching MP2 dividend rates.");
    throw error;
  }
};

export const updateMp2Rates = async (rates: Mp2DividendRates): Promise<Mp2DividendRates> => {
  try {
    return await invoke<Mp2DividendRates>("update_mp2_rates", { rates });
  } catch (error) {
    logger.error("Error updating MP2 dividend rates.");
    throw error;
  }
};
